File: internal/watcher/watcher.go
// Package watcher monitors configured source paths for changes and triggers
// the appropriate parser + generator pipeline on each modification event.
package watcher

import (
	"log"
	"path/filepath"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
	"github.com/inevolin/histd/internal/config"
	"github.com/inevolin/histd/internal/generator"
	"github.com/inevolin/histd/internal/parser"
)

const debounceDuration = 2 * time.Second

// parsers is the ordered list of registered HistoryParsers.  The first parser
// whose CanHandle returns true is used for a given file.
var parsers = []parser.HistoryParser{
	&parser.ClaudeParser{},
	&parser.CursorParser{},
}

// Run starts the file-watching loop.  It blocks until an error occurs on the
// watcher itself.  Each watched path from cfg is added recursively if it is a
// directory.
func Run(cfg *config.Config) error {
	w, err := fsnotify.NewWatcher()
	if err != nil {
		return err
	}
	defer w.Close()

	for _, entry := range cfg.Watches {
		if err := addPath(w, entry.Path); err != nil {
			log.Printf("watcher: cannot watch %s: %v", entry.Path, err)
		} else {
			log.Printf("watcher: watching %s (%s)", entry.Path, entry.Tool)
		}
	}

	debounce := newDebouncer()

	for {
		select {
		case event, ok := <-w.Events:
			if !ok {
				return nil
			}
			if event.Has(fsnotify.Write) || event.Has(fsnotify.Create) {
				path := event.Name
				debounce.trigger(path, func() {
					handleChange(path, cfg.OutputDir)
				})
			}
		case err, ok := <-w.Errors:
			if !ok {
				return nil
			}
			log.Printf("watcher: error: %v", err)
		}
	}
}

// addPath adds a single path (file or directory) to the watcher.
func addPath(w *fsnotify.Watcher, path string) error {
	return w.Add(path)
}

// handleChange runs the parser + generator pipeline for a changed file.
func handleChange(path, outputDir string) {
	p := selectParser(path)
	if p == nil {
		return
	}
	sessions, err := p.Parse(path)
	if err != nil {
		log.Printf("watcher: parse error for %s: %v", path, err)
		return
	}
	if len(sessions) == 0 {
		return
	}
	if err := generator.Write(sessions, outputDir); err != nil {
		log.Printf("watcher: generator error for %s: %v", path, err)
		return
	}
	log.Printf("watcher: wrote %d session(s) from %s", len(sessions), filepath.Base(path))
}

// selectParser returns the first registered parser that can handle path.
func selectParser(path string) parser.HistoryParser {
	for _, p := range parsers {
		if p.CanHandle(path) {
			return p
		}
	}
	return nil
}

// debouncer coalesces rapid events for the same path into a single action
// that fires debounceDuration after the last event.  A version counter ensures
// that if a timer fires just as a new trigger arrives, the stale callback is
// discarded and fn is called only once per quiet window.
type debouncer struct {
	mu      sync.Mutex
	timers  map[string]*time.Timer
	version map[string]int
}

func newDebouncer() *debouncer {
	return &debouncer{
		timers:  make(map[string]*time.Timer),
		version: make(map[string]int),
	}
}

func (d *debouncer) trigger(key string, fn func()) {
	d.mu.Lock()
	defer d.mu.Unlock()
	if t, ok := d.timers[key]; ok {
		t.Stop()
	}
	d.version[key]++
	v := d.version[key]
	d.timers[key] = time.AfterFunc(debounceDuration, func() {
		d.mu.Lock()
		if d.version[key] != v {
			// A newer trigger replaced us; skip this invocation.
			d.mu.Unlock()
			return
		}
		delete(d.timers, key)
		delete(d.version, key)
		d.mu.Unlock()
		fn()
	})
}

File: src/watcher.ts
import * as path from 'path';
import { Config } from './config.js';
import { write } from './generator.js';
import { HistoryParser } from './parser/types.js';
import { ClaudeParser } from './parser/claude.js';
import { CursorParser } from './parser/cursor.js';

const DEBOUNCE_MS = 2000;

/** Ordered list of registered HistoryParsers. */
const parsers: HistoryParser[] = [
  new ClaudeParser(),
  new CursorParser(),
];

/**
 * Placeholder: file-watching is replaced by the MCP server architecture.
 * This module will be deleted in a subsequent task.
 */
export function run(_cfg: Config): Promise<void> {
  return Promise.reject(new Error('watcher is no longer supported; use the MCP server'));
}

/** Runs the parser + generator pipeline for a changed file. */
async function handleChange(filePath: string, outputDir: string): Promise<void> {
  const p = selectParser(filePath);
  if (!p) return;

  try {
    const sessions = await p.parse(filePath);
    if (sessions.length === 0) return;
    write(sessions, outputDir);
    console.log(`watcher: wrote ${sessions.length} session(s) from ${path.basename(filePath)}`);
  } catch (err) {
    console.error(`watcher: error processing ${filePath}: ${err}`);
  }
}

/** Returns the first registered parser that can handle filePath. */
function selectParser(filePath: string): HistoryParser | null {
  for (const p of parsers) {
    if (p.canHandle(filePath)) return p;
  }
  return null;
}

/**
 * Coalesces rapid events for the same path into a single action that fires
 * DEBOUNCE_MS after the last event. A version counter ensures that stale
 * callbacks are discarded.
 */
class Debouncer {
  private timers = new Map<string, ReturnType<typeof setTimeout>>();
  private versions = new Map<string, number>();

  trigger(key: string, fn: () => void): void {
    const existing = this.timers.get(key);
    if (existing !== undefined) {
      clearTimeout(existing);
    }

    const v = (this.versions.get(key) ?? 0) + 1;
    this.versions.set(key, v);

    const timer = setTimeout(() => {
      if (this.versions.get(key) !== v) {
        // A newer trigger replaced us; skip this invocation.
        return;
      }
      this.timers.delete(key);
      this.versions.delete(key);
      fn();
    }, DEBOUNCE_MS);

    this.timers.set(key, timer);
  }
}
